### PR TITLE
maturin 0.10.6 (new formula)

### PR DESCRIPTION
**Pull request by messense**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/maturin.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  head "https://github.com/PyO3/maturin.git"
+
+  depends_on "rust" => :build
+
+  def install
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    assert_match "Build and publish crates with pyo3", shell_output("#{bin}/maturin --help 2>&1")
```
Comment (carlocab) [inline Formula/maturin.rb:16]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

Comment (messense) [reply Formula/maturin.rb:16]: Updated.

File: Formula/maturin.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    system "cargo", "new", "hello_world", "--bin"
+    system "#{bin}/maturin", "build", "-m", "hello_world/Cargo.toml", "-b", "bin", "-o", "dist"
```
Comment (carlocab) [inline Formula/maturin.rb:19]: This is better. Can we test that this created a Python package correctly? We can probably use `assert_predicate` with `:exist?` and `assert_match` to read substrings from any generated files.

File: Formula/maturin.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class Maturin < Formula
+  desc "Build and publish Rust crates as Python packages"
+  homepage "https://github.com/PyO3/maturin"
+  url "https://github.com/PyO3/maturin/archive/refs/tags/v0.10.6.tar.gz"
+  sha256 "e6a9a67cc62ffe248654e60e7ec211bf23319c4c936ad87022f7a1fd0997430d"
+  license "MIT"
+  head "https://github.com/PyO3/maturin.git"
+
+  depends_on "rust" => :build
```
Comment (gromgit) [inline Formula/maturin.rb:9]:
```suggestion
  depends_on "rust"
```
Based on the new test, I assume `maturin` requires Rust at runtime.

File: Formula/maturin.rb (hunk 2 above)
Comment (messense) [reply Formula/maturin.rb:19]: Not sure how to use `assert_match` to read a `.whl` file. (A zip file)

Comment (messense) [reply Formula/maturin.rb:18]: Update to test package installation with python3.

File: Formula/maturin.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  depends_on "rust"
+
+  def install
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    system "cargo", "new", "hello_world", "--bin"
+    system "#{bin}/maturin", "build", "-m", "hello_world/Cargo.toml", "-b", "bin", "-o", "dist"
+    assert_predicate testpath/"dist/hello_world-0.1.0-py3-none-macosx_10_7_x86_64.whl", :exist?,
```
Comment (messense) [inline Formula/maturin.rb:19]: Is there a variable for current system architecture? `x86_64` will fail on M1 macOS.

File: Formula/maturin.rb (hunk 2 above)
Comment (carlocab) [reply Formula/maturin.rb:18]: The new test looks good. Thanks for your work on this.

File: Formula/maturin.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+class Maturin < Formula
+  desc "Build and publish Rust crates as Python packages"
+  homepage "https://github.com/PyO3/maturin"
+  url "https://github.com/PyO3/maturin/archive/refs/tags/v0.10.6.tar.gz"
+  sha256 "e6a9a67cc62ffe248654e60e7ec211bf23319c4c936ad87022f7a1fd0997430d"
+  license "MIT"
+  head "https://github.com/PyO3/maturin.git"
```
Comment (nandahkrishna) [inline Formula/maturin.rb:7]:
```suggestion
  head "https://github.com/PyO3/maturin.git", branch: "main"
```

**Review by nandahkrishna — COMMENTED** [review]
Minor suggestion, the repo's default branch is `main` so `HEAD` installs won't work unless we specify it here.

**Review by nandahkrishna — COMMENTED** [review]
Looks good to me, thanks @messense!

**Review by carlocab — APPROVED** [review]
Thanks for your contribution, @messense! And thanks for following our contribution guidelines carefully.

This can be merged when CI is done.

**Pull request closed without merging**